File: 2491-divide-players-into-teams-of-equal-skill/2491-divide-players-into-teams-of-equal-skill.cpp
class Solution {
public:
    long long dividePlayers(vector<int>& skill) {
        long long sum = 0, ans = 0;
        for(int s:skill) sum += s;
        int equal = skill.size()/2;
        if(sum%equal != 0) return -1;
        int k = sum / equal;
        unordered_map<int, int>mp;
        for(int x:skill) {
            if(mp[k - x] > 0) {
                mp[k - x]--;
                ans += x * (k - x);
            }
            else {
                mp[x]++;
            }
        }
        for(auto it:mp) {
            if(it.second > 0) return -1;
        }
        return ans;
    }
};

File: 2491-divide-players-into-teams-of-equal-skill/2491-divide-players-into-teams-of-equal-skill.java
class Solution {
    public long dividePlayers(int[] skill) {
        long ans = 0, sum = 0;
        for(int x:skill) {
            sum += x;
        }
        long equal;
        equal = sum / (long)(skill.length/2);
        if(sum%equal != 0) return -1;
        int k = (int)equal; 
        Map<Integer, Integer>mp = new HashMap<>();
        for(int x:skill) {
            if(mp.containsKey(k - x) && mp.get(k-x) > 0) {
                mp.put(k-x, mp.get(k-x) - 1);
                ans += (long)x*(k-x);
            }
            else {
                mp.put(x, mp.getOrDefault(x, 0) + 1);
            }
        }
        for(Map.Entry<Integer, Integer> entry: mp.entrySet()) {
            if(entry.getValue() > 0) return -1;
        }
        return ans;
    }
}